Kubernetes: Merge StatefulSet resources from input if set

Sgitario opened the pull request:
When users add a StatefulSet resource and the deployment-target is set to `StatefulSet`, we need to read the StatefulSet resource from the user and populate it accordingly.

Before these changes, we were including the StatefulSet resource, but adding a new one (with the same name), so in the end, two StatefulSet resources were included in the final kubernetes.yml.

Fix https://github.com/quarkusio/quarkus/issues/25162

Sgitario commented:
cc @geoand @iocanel @tarilabs

iocanel commented on extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/AddStatefulSetResourceDecorator.java:
```diff
@@ -1,43 +1,80 @@
+                        .endMetadata());
+
+        // init main sub items
+        StatefulSetFluent.SpecNested<StatefulSetBuilder> specBuilder = builder.editOrNewSpec()
```
This is not a change request but more of a suggestion:

```java
        StatefulSet statefulSet = list.getItems().stream()
                .filter(i -> Strings.isNotNullOrEmpty(i.getKind()) && i.getKind().equalsIgnoreCase(STATEFULSET))
                .filter(i -> i.getMetadata().getName().equals(name))
                .map(metadata -> {
                    list.removeFromItems(metadata);
                    return new StatefulSetBuilder((StatefulSet) metadata);
                }).findAny()
                .orElseGet(() -> new StatefulSetBuilder()
                        .withNewMetadata()
                        .withName(name)
                        .endMetadata())
               .accept(StatefulSetSpecBuilder.class, spec -> {
                        if (spec.getReplicas() != null) {
                            spec.withReplicas(1);
                        }
                        if (Strings.isNullOrEmpty(spec.getServiceName())) {
                            spec.withServiceName(name);
                        }
                        if (spec.getSelector().getMatchLabels() == null) {
                            spec.editSelector().withMatchLabels(new HashMap<>()).endSelector();
                        }
                        // - termination grace period seconds
                        if (spec.getTemplate().getSpec().getTerminationGracePeriodSeconds() == null) {
                            spec.editTemplate().editSpec().withTerminationGracePeriodSeconds(10L).endSpec().endTemplate();
                        }
                        //containers
                        if (!containsContainerWithName(spec)) {
                           spec.editTemplate()
                                  .editSpec()
                                  .addNewContainer()
                                  .withName(name)
                                  .endContainer()
                                  .endSpec().endTemplate();
                            } 
                         }).build(); 
```

This mostly avoid referring to internal / nested builder types and gives a more fluent feel.

Sgitario replied:
I've refactored this method according to your suggestion.

iocanel commented on integration-tests/kubernetes/quarkus-standard-way/src/test/java/io/quarkus/it/kubernetes/KubernetesWithInputStatefulSetResourcesTest.java:
```diff
@@ -0,0 +1,78 @@
+            .withApplicationRoot((jar) -> jar.addClasses(GreetingResource.class))
+            .setApplicationName(APP_NAME)
+            .setApplicationVersion("0.1-SNAPSHOT")
+            .withConfigurationResource("kubernetes-with-statefulset-resource.properties")
```
Lately, I find that embedding the actual configuration in the file makes the test code easier to maintain.
Again, this is styling thing and not a change request.

Sgitario replied:
I tried to add the properties from the `withApplicationRoot` and for some reason, didn't work for me. But I do agree with you that this change will improve this test and the rest of the tests within this module. Let's do this for all of them in one go in a later pr.

iocanel reviewed (approved):
LGTM

Added some some comments, that are not really change requests but more like optional suggestions that you may or may not find useful.

Sgitario committed:
Kubernetes: Merge StatefulSet resources from input if set

When users add a StatefulSet resource and the deployment-target is set to `StatefulSet`, we need to read the StatefulSet resource from the user and populate it accordingly.

Before these changes, we were including the StatefulSet resource, but adding a new one (with the same name), so in the end, two StatefulSet resources were included in the final kubernetes.yml.

Fix https://github.com/quarkusio/quarkus/issues/25162
